# RNA-seqæµç¨‹

# 1.å‰æœŸå‡†å¤‡

åœ¨cç›˜å»ºç«‹æ–‡ä»¶å¤¹
```
cd /mnt/c
mkdir -p project/rat
cd project/rat
mkdir annotation genome sequence output script

tree
â”œâ”€â”€ annotation
â”œâ”€â”€ genome
â”œâ”€â”€ output
â”œâ”€â”€ script
â””â”€â”€ sequence
```
# 2.å·¥å…·ä¸‹è½½
## 2.1 sratoolkit
```
#å®‰è£…
brew reinstall sratoolkit

#éªŒè¯æ˜¯å¦å®‰è£…æˆåŠŸ
prefetch --help
```
* fastq-dump  

SRA Toolkitä¸­æœ€å¸¸ç”¨çš„å·¥å…·ï¼Œä¸»è¦åŠŸèƒ½æ˜¯å°†SRAæ•°æ®è½¬æ¢ä¸ºfastqæ ¼å¼ã€‚ç›¸å½“äºè§£å‹ç¼©  

ç”¨æ³•ï¼š
```
fastq-dump [options] <path/file>
```
* prefetch  

ä½¿ç”¨æ–¹æ³•:
```
 prefetch [options] <SRA accession> [...]
  Download SRA files and their dependencies

  prefetch [options] --cart <kart file>
  Download cart file

  prefetch [options] <URL> --output-file <FILE>
  Download URL to FILE

  prefetch [options] <URL> [...] --output-directory <DIRECTORY>
  Download URL or URL-s to DIRECTORY

  prefetch [options] <SRA file> [...]
  Check SRA file for missed dependencies and download them

```
## 2.2 fastqc

```
#å®‰è£…
brew reinstall fastqc

#éªŒè¯æ˜¯å¦å®‰è£…æˆåŠŸ
$ fastqc -v
FastQC v0.12.1
```
## 2.3 multiqc
```
# ä½¿ç”¨pythonçš„å®‰è£…å™¨å®‰è£…
pip install 

#éªŒè¯æ˜¯å¦å®‰è£…æˆåŠŸ
multiqc --help
```
## 2.4 cutadapt
```
#å®‰è£…
pip install cutadapt 

#éªŒè¯æ˜¯å¦å®‰è£…æˆåŠŸ
$ cutadapt --version
4.4
```
## 2.5 è´¨é‡ä¿®å‰ª
* Trim Galore
```
## éœ€å…ˆå®‰è£…fastqcå’Œcutadaptå†å®‰è£…trim_galoreï¼Œè¦å®‰è£…æ–°ç‰ˆçš„trim_galoreï¼Œæ—§ç‰ˆçš„æŠ¥é”™
cd /mnt/c/project/biosoft
wget https://github.com/FelixKrueger/TrimGalore/archive/0.6.10.tar.gz -o trim_galore.tar.gz
tar zxvf 0.4.5.tar.gz

# åœ¨ç³»ç»Ÿä¸­é…ç½®ç¯å¢ƒå˜é‡
vim ~/.bashrc
åœ¨ç•Œé¢ä¸­è¾“å…¥
# trim_galore
export PATH="/mnt/c/project/biosoft/TrimGalore-0.4.5:$PATH"
#ä¿å­˜åé€€å‡ºå¹¶æ›´æ–°
source ~/.bashrc

#éªŒè¯æ˜¯å¦å®‰è£…æˆåŠŸ
trim_galore -v
Quality-/Adapter-/RRBS-/Speciality-Trimming
                                [powered by Cutadapt]
                                  version 0.6.10

                               Last update: 02 02 2023
```
* trimmomatic
```
cd /mnt/c/project/biosoft
wget http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.38.zip
unzip Trimmomatic-0.38.zip
cd Trimmomatic-0.38

# å¯¼å…¥ä¸´æ—¶ç¯å¢ƒå˜é‡
export PATH="$(pwd):$PATH"
```



## 2.6 hisat2
```
# ä¸‹è½½å®‰è£…
wget https://cloud.biohpc.swmed.edu/index.php/s/oTtGWbWjaxsQ2Ho/download

# åœ¨ç³»ç»Ÿä¸­é…ç½®ç¯å¢ƒå˜é‡
vim ~/.bashrc
åœ¨ç•Œé¢ä¸­è¾“å…¥
# hisat2
export PATH="/mnt/c/project/biosoft/hisat2-2.2.1:$PATH"
# åˆ·æ–°
source ~/.bashrc

# æµ‹è¯•æ˜¯å¦å¯ç”¨
$ hisat2 -h
```
## 2.7 samtools
```
#å®‰è£…
brew install samtools
```
## 2.8 HTseq
```
#å®‰è£…
pip install -i https://pypi.tuna.tsinghua.edu.cn/simple HTseq
```
## 2.9 R
```
brew install r
```
ç›´æ¥åœ¨å®˜ç½‘é€‰æ‹©``Download R for Windows``; ``install R for the first time``ä¸‹è½½å®‰è£…åŒ…å³å¯ï¼›æ³¨æ„å¯ä»¥å°†ä¸¤ä¸ªæ–‡ä»¶æ”¾åœ¨åŒä¸€ä¸ªæ–‡ä»¶å¤¹å†…
## 2.10 Rstudio
è¿›å…¥ç½‘ç«™ï¼šhttps://www.rstudio.com/products/rstudio/download/
R studio å¯ä»¥åœ¨ Windows ä¸‹å®‰è£…; é€‰æ‹©ç‰ˆæœ¬ä¸‹è½½,ä¸‹è½½å®Œæˆä¹‹ååŒå‡»å®‰è£…ã€‚
## 2.11 parallel
```
brew install parallel
```
## 3. æ•°æ®ä¸‹è½½
### 3.1 åŸºå› ç»„æ•°æ®
1. [ensembleç½‘å€](https://asia.ensembl.org/)ï¼Œåœ¨å·¦ä¾§äº”ä¸­ä¸­é€‰æ‹©``Rat``; åœ¨å·¦ä¾§Download DNA sequence (FASTA) ä¸‹è½½åŸºå› ç»„åºåˆ—æ•°æ®; åœ¨å³ä¾§çš„Download GTF or GFF3 (files for genes, cDNAs, ncRNA, proteins)ä¸‹è½½åŸºå› æ³¨é‡Šæ–‡ä»¶ã€‚

å¯ä»¥ç›´æ¥åœ¨ç½‘é¡µä¸‹è½½ï¼Œä¹Ÿå¯ç”¨ä»£ç 
```
# ä¸‹è½½
cd /mnt/d/project/rat/genome
wget http://ftp.ensembl.org/pub/release-107/fasta/rattus_norvegicus/dna/Rattus_norvegicus.mRatBN7.2.dna.toplevel.fa.gz
gzip -d Rattus_norvegicus.mRatBN7.2.dna.toplevel.fa.gz

# æ”¹åï¼ˆæ–¹ä¾¿åé¢ä½¿ç”¨ï¼Œåå­—å¤ªé•¿ä¸€æ¥ä¸æ–¹ä¾¿è¾“å…¥ï¼ŒäºŒæ¥å¯èƒ½ä¼šè¾“é”™ï¼‰
mv Rattus_norvegicus.mRatBN7.2.dna.toplevel.fa mRatBN7.2.fa

```
ä¸‹è½½å¾—åˆ°çš„åŸºå› ç»„æ–‡ä»¶å¯ä»¥æŸ¥çœ‹ä¸€ä¸‹åŒ…å«å“ªäº›æŸ“è‰²ä½“ï¼Œç¡®è®¤æ–‡ä»¶æ˜¯å¦ä¸‹è½½æ­£ç¡®ã€‚
```
cat mRatBN7.2.fa | grep "^>"
```
ç»“æœï¼šé™¤äº†1-20å·+X+Y+MTä¹‹å¤–è¿˜æœ‰å¾ˆå¤šåˆ«çš„IDåã€‚è¿™äº›éƒ½æ˜¯scaffold
```
>19 dna:primary_assembly primary_assembly:mRatBN7.2:19:1:57337602:1 REF
>20 dna:primary_assembly primary_assembly:mRatBN7.2:20:1:54435887:1 REF
>X dna:primary_assembly primary_assembly:mRatBN7.2:X:1:152453651:1 REF
>Y dna:primary_assembly primary_assembly:mRatBN7.2:Y:1:18315841:1 REF
>MT dna:primary_assembly primary_assembly:mRatBN7.2:MT:1:16313:1 REF
>MU150191.1 dna:primary_assembly primary_assembly:mRatBN7.2:MU150191.1:1:1794995:1 REF

>JACYVU010000493.1 dna:primary_assembly primary_assembly:mRatBN7.2:JACYVU010000493.1:1:444596:1 REF
>MU150193.1 dna:primary_assembly primary_assembly:mRatBN7.2:MU150193.1:1:383091:1 REF
```
æ¯ä¸€æ¡primary_assemblyçš„åç§°åé¢è¿˜è·Ÿäº†ä¸€äº›æè¿°ä¿¡æ¯ï¼Œè¿™äº›æè¿°ä¿¡æ¯å°±æ˜¯å½“å‰ç»„è£…ç‰ˆæœ¬ï¼Œé•¿åº¦ç­‰ç­‰ä¿¡æ¯ï¼Œä½†æ˜¯è¿™ä¸ªä¿¡æ¯ä¼šå¦¨ç¢åé¢å†™è„šæœ¬ç»Ÿè®¡æˆ–è€…ä¸€äº›åˆ†æï¼Œæ‰€ä»¥è¿™é‡Œæœ€å¥½å»æ‰
```
# é¦–å…ˆå°†ä¹‹å‰çš„åç§°æ›´æ”¹ä¸€ä¸‹
mv mRatBN7.2.fa mRatBN7.2.raw.fa

# ç„¶åå»é™¤æŸ“è‰²ä½“ç¼–å·åçš„æè¿°ä¿¡æ¯
$ cat mRatBN7.2.raw.fa | perl -n -e 'if(m/^>(.+?)(?:\s|$)/){ print ">$1\n";}else{print}' > mRatBN7.2.fa
#å•è¡ŒåŒ¹é…ï¼Œå¦‚æœåŒ¹é…åˆ°äº† å¼€å¤´>å¤šä¸ªå­—æ¯ç©ºæ ¼æˆ–è€…$ï¼Œå°†  >å¤šä¸ªå­—æ¯  æ‰“å°å‡ºæ¥

# åˆ é™¤
$ rm mRatBN7.2.raw.fa
```
### 3.2 ä¸‹è½½æ³¨é‡Šä¿¡æ¯
```
# ä¸‹è½½ gff3 æ ¼å¼
cd /mnt/c/project/rat/annotation
wget http://ftp.ensembl.org/pub/release-107/gff3/rattus_norvegicus/Rattus_norvegicus.mRatBN7.2.107.gff3.gz
gzip -d Rattus_norvegicus.mRatBN7.2.107.gff3.gz
# åŒæ ·çš„ä¹Ÿæ”¹å
mv Rattus_norvegicus.mRatBN7.2.107.gff3 mRatBN7.2.gff
# ä½¿ç”¨headæŸ¥çœ‹éƒ¨åˆ†
head mRatBN7.2.gff
```

```
# ä¸‹è½½ gtf æ ¼å¼  
cd /mnt/c/project/rat/annotation
wget http://ftp.ensembl.org/pub/release-107/gtf/rattus_norvegicus/Rattus_norvegicus.mRatBN7.2.107.gtf.gz
gzip -d Rattus_norvegicus.mRatBN7.2.107.gtf.gz
```
### 3.3 ä¸‹è½½RNA-SEQæ•°æ®
```
cd /mnt/c/project/rat/sequence
prefetch SRR2190795 SRR224018{2..7} SRR2240228

#æˆ–è€…è¿™æ ·ä¹Ÿå¯ä»¥
cat >1.txt <<EOF
SRR2190795
SRR2240182
SRR2240183
SRR2240184
SRR2240185
SRR2240186
SRR2240187
SRR2240228
EOF
prefetch --option-file 1.txt
```

æ ¼å¼è½¬æ¢

ä¸‹è½½å¾—åˆ°.sraæ–‡ä»¶ï¼Œä½¿ç”¨SRAtoolkitå·¥å…·åŒ…çš„fastq-dumpå·¥å…·ï¼Œä½¿ç”¨å®ƒæ¥è¿›è¡Œæ ¼å¼è½¬åŒ–
```
ls
SRR2190795.sra  SRR2240183.sra  SRR2240185.sra  SRR2240187.sra
SRR2240182.sra  SRR2240184.sra  SRR2240186.sra  SRR2240228.sra

$ parallel -j 4 "    # ç”¨parallelå¤šçº¿ç¨‹åŠ å¿«é€Ÿåº¦ï¼Œå¹¶è¡Œä»»åŠ¡æ•°ä¸º4
    fastq-dump --split-3 --gzip {1}    # å°†sraæ–‡ä»¶è½¬åŒ–ä¸ºfastqæ–‡ä»¶ä¹‹åå‹ç¼©ä¸ºgzæ–‡ä»¶
" ::: $(ls *.sra)     # :::åæ¥å¯¹è±¡

fastq-dump --split-3 --gzip SRR7368844.sra

# ls *.sraä»£è¡¨ï¼Œåˆ—ä¸¾å‡ºä»»ä½•ä»¥.sraç»“å°¾çš„æ–‡ä»¶
--gzip å°†è½¬æ¢å‡ºçš„fastqæ–‡ä»¶ä»¥gzæ ¼å¼è¾“å‡ºï¼Œå¯ä»¥èŠ‚çœç©ºé—´
--split-3 æŠŠpair-endæµ‹åºåˆ†æˆä¸¤ä¸ªæ–‡ä»¶è¾“å‡ºï¼Œå¯ç”¨äºåŒç«¯æµ‹åºè½¬åŒ–ä¸ºä¸¤ä¸ªæ–‡ä»¶ï¼Œæœ¬æ–‡ä¸¾ä¾‹ä¸ºå•ç«¯æµ‹åºï¼Œåˆ æ‰ä¸å½±å“
-O è¾“å‡ºæ–‡ä»¶å¤¹åï¼Œä¸åŠ ç›´æ¥æ”¾åœ¨è¯¥æ–‡ä»¶å¤¹

# åˆ é™¤sraæ–‡ä»¶
$ rm *.sra

```
æ ¼å¼ä»‹ç»
```
# æŸ¥çœ‹ä¸‹è½½å¥½çš„gzæ–‡ä»¶
   cd ~/data/sra
   gzip -d -c SRR2190795.fastq.gz | head -n 20

# gzip
-cæˆ–--stdoutæˆ–--to-stdout ã€€æŠŠå‹ç¼©åçš„æ–‡ä»¶è¾“å‡ºåˆ°æ ‡å‡†è¾“å‡ºè®¾å¤‡ï¼Œä¸å»æ›´åŠ¨åŸå§‹æ–‡ä»¶ã€‚
-dæˆ–--decompressæˆ–----uncompress ã€€è§£å¼€å‹ç¼©æ–‡ä»¶ã€‚

```
ç»“æœï¼š
```
@SRR2190795.1 HWI-ST1147:240:C5NY7ACXX:1:1101:1320:2244 length=100
ATGCTGGGGGCATTAGCATTGGGTACTGAATTATTTTCAGTAAGAGGGAAAGAATCCATCTCCNNNNNNNNNNNNNNNNNNNNNNAAANAAAAATAAAAT
+SRR2190795.1 HWI-ST1147:240:C5NY7ACXX:1:1101:1320:2244 length=100
CCCFFFFFHHHHHJIJJJJJJJJDHHJJJIJJJJJIJJJJJJJJJJJJJJJJJJJJJJJJJHH#####################################
@SRR2190795.2 HWI-ST1147:240:C5NY7ACXX:1:1101:1598:2247 length=100
AACTTCGGTTCTCTACTAGGAGTATGCCTCATAGTACAAATCCTCACAGGCTTATTCCTAGCANNNNNNNNNNNNNNNNNNNNNNTAACAGCATTTTCAT
+SRR2190795.2 HWI-ST1147:240:C5NY7ACXX:1:1101:1598:2247 length=100
@@@7D8+@A:1CFG<C:23<:E<;FF<BHIIEHG:?:??CDF<9DCGGG?1?FEG@@<@CA#######################################
@SRR2190795.3 HWI-ST1147:240:C5NY7ACXX:1:1101:1641:2250 length=100
AGAAGGTCTTAGATCAGAAGGAGCACAGACTGGATGGTCGTGTCATTGACCCTAAAAAGGCTANNNNNNNNNNNNNNNNNNNNNTGAAGAAAATCTTTGT
+SRR2190795.3 HWI-ST1147:240:C5NY7ACXX:1:1101:1641:2250 length=100
BC@FFFDDHHHHHJJJJJJJJJJJJJJJJJJJJIJJJFHGHHEGHIIIHJIJJIJJIJIJJID#####################################
@SRR2190795.4 HWI-ST1147:240:C5NY7ACXX:1:1101:1851:2233 length=100
GGGATTTCATGGCCTCCACGTAATTATTGGCTCAACTTTCCTAATTGTCTGTCTACTACGACANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTNNCNNN
+SRR2190795.4 HWI-ST1147:240:C5NY7ACXX:1:1101:1851:2233 length=100
@@?DDBDDFFDDDGHGGGGI?B;FFHGHA@FEHGHDDGHEGGFGHIGEHIIHIGGBGACD6AH#####################################
@SRR2190795.5 HWI-ST1147:240:C5NY7ACXX:1:1101:1957:2243 length=100
CAGCCATTGTGGCTCCCGATGGCTTTGACATCATTGACATGACAGCCGGAGGTCAGATAAACTNNNNNNNNNNNNNNNNNNNNNNATCNGTGGCAAAGGT
+SRR2190795.5 HWI-ST1147:240:C5NY7ACXX:1:1101:1957:2243 length=100
@CCFFFFFHHHHAHJJJIJJJJJJIJJIGGGIFIJIIHIIGGJJJJJJJFHIGIJHHHHHHFC#####################################
```
## 4. è´¨é‡æ§åˆ¶
### 4.1ç”¨ fastqc è¿›è¡Œè´¨é‡è¯„ä¼°
1. å…·ä½“ç”¨æ³•
```
# åŸºæœ¬æ ¼å¼

# fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] [-c contaminant file] seqfile1 .. seqfileN

# ä¸»è¦æ˜¯åŒ…æ‹¬å‰é¢çš„å„ç§é€‰é¡¹å’Œæœ€åé¢çš„å¯ä»¥åŠ å…¥Nä¸ªæ–‡ä»¶

# -o --outdir FastQCç”Ÿæˆçš„æŠ¥å‘Šæ–‡ä»¶çš„å‚¨å­˜è·¯å¾„ï¼Œç”Ÿæˆçš„æŠ¥å‘Šçš„æ–‡ä»¶åæ˜¯æ ¹æ®è¾“å…¥æ¥å®šçš„

# --extract ç”Ÿæˆçš„æŠ¥å‘Šé»˜è®¤ä¼šæ‰“åŒ…æˆ1ä¸ªå‹ç¼©æ–‡ä»¶ï¼Œä½¿ç”¨è¿™ä¸ªå‚æ•°æ˜¯è®©ç¨‹åºä¸æ‰“åŒ…

# -t --threads é€‰æ‹©ç¨‹åºè¿è¡Œçš„çº¿ç¨‹æ•°ï¼Œæ¯ä¸ªçº¿ç¨‹ä¼šå ç”¨250MBå†…å­˜ï¼Œè¶Šå¤šè¶Šå¿«å’¯

# -c --contaminants æ±¡æŸ“ç‰©é€‰é¡¹ï¼Œè¾“å…¥çš„æ˜¯ä¸€ä¸ªæ–‡ä»¶ï¼Œæ ¼å¼æ˜¯Name [Tab] Sequenceï¼Œé‡Œé¢æ˜¯å¯èƒ½çš„æ±¡æŸ“åºåˆ—ï¼Œå¦‚æœæœ‰è¿™ä¸ªé€‰é¡¹ï¼ŒFastQCä¼šåœ¨è®¡ç®—æ—¶å€™è¯„ä¼°æ±¡æŸ“çš„æƒ…å†µï¼Œå¹¶åœ¨ç»Ÿè®¡çš„æ—¶å€™è¿›è¡Œåˆ†æï¼Œä¸€èˆ¬ç”¨ä¸åˆ°

# -a --adapters ä¹Ÿæ˜¯è¾“å…¥ä¸€ä¸ªæ–‡ä»¶ï¼Œæ–‡ä»¶çš„æ ¼å¼Name [Tab] Sequenceï¼Œå‚¨å­˜çš„æ˜¯æµ‹åºçš„adpateråºåˆ—ä¿¡æ¯ï¼Œå¦‚æœä¸è¾“å…¥ï¼Œç›®å‰ç‰ˆæœ¬çš„FastQCå°±æŒ‰ç…§é€šç”¨å¼•ç‰©æ¥è¯„ä¼°åºåˆ—æ—¶å€™æœ‰adapterçš„æ®‹ç•™

# -q --quiet å®‰é™è¿è¡Œæ¨¡å¼ï¼Œä¸€èˆ¬ä¸é€‰è¿™ä¸ªé€‰é¡¹çš„æ—¶å€™ï¼Œç¨‹åºä¼šå®æ—¶æŠ¥å‘Šè¿è¡Œçš„çŠ¶å†µã€‚
```
2. è¾“å…¥ä»£ç 
```
# æ–°å»ºç›®å½•  
mkdir /mnt/c/project/rat/output/fastqc

ï¼æ³¨æ„ï¼ä¸€å®šåœ¨å­˜å‚¨fastqc.gzçš„æ–‡ä»¶å¤¹è·¯å¾„ä¸‹æ‰§è¡Œä¸‹é¢çš„å‘½ä»¤
$ cd /mnt/c/project/rat/output/sequence

fastqc -t 6 -o /mnt/c/project/rat/output/fastqc *.gz
# -t æŒ‡å®šçº¿ç¨‹æ•°
# -o æŒ‡å®šè¾“å‡ºæ–‡ä»¶å¤¹
# *.gz è¡¨ç¤ºè¿™ä¸ªç›®å½•ä¸‹ä»¥ .gz çš„æ‰€æœ‰æ–‡ä»¶
```
3. ç»“æœ
```
cd /mnt/c/project/rat/output/fastqc
ls
SRR2190795_fastqc.html  SRR2240183_fastqc.html  SRR2240185_fastqc.html  SRR2240187_fastqc.html
SRR2190795_fastqc.zip   SRR2240183_fastqc.zip   SRR2240185_fastqc.zip   SRR2240187_fastqc.zip
SRR2240182_fastqc.html  SRR2240184_fastqc.html  SRR2240186_fastqc.html  SRR2240228_fastqc.html
SRR2240182_fastqc.zip   SRR2240184_fastqc.zip   SRR2240186_fastqc.zip   SRR2240228_fastqc.zip
```
4. åˆ†æ

.htmlç”¨æµè§ˆå™¨æ‰“å¼€ï¼ŒæŸ¥çœ‹æƒ…å†µ

5. å°†æ‰€æœ‰çš„fastqcçš„æ£€æµ‹æŠ¥å‘Šåˆå¹¶åˆ°ä¸€ä¸ªæ–‡ä»¶ä¸Šçš„ç¨‹åºmultiqcï¼Œæ–¹ä¾¿åˆ†æ
```
cd /mnt/c/project/rat/output/fastqc

multiqc .

/// MultiQC ğŸ” | v1.12

|           multiqc | Search path : /mnt/d/project/rat/output/fastqc
|         searching | â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â” 100% 16/16
|            fastqc | Found 8 reports
|           multiqc | Compressing plot data
|           multiqc | Report      : multiqc_report.html
|           multiqc | Data        : multiqc_data
|           multiqc | MultiQC complete

ç»“æœ: åœ¨è¯¥æ–‡ä»¶å¤¹å†…å¾—åˆ°ä¸€ä¸ªmultiqc_report.htmlæ–‡ä»¶
```
## 4.2 å‰”é™¤æ¥å¤´ä»¥åŠæµ‹åºè´¨é‡å·®çš„ç¢±åŸº

é‡‡ç”¨trimmomaticè¿›è¡Œä¸¤ç«¯ä½è´¨é‡çš„åŒºåŸŸçš„å»é™¤
```
# æ–°å»ºæ–‡ä»¶å¤¹
$ mkdir -p ../output/adapter

cd /mnt/c/project/rat/sequence

for i in $(ls *.fastq.gz);
do
cutadapt -a AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT \
    --minimum-length 30 --overlap 4 --trim-n \
    -o /mnt/c/project/rat/output/adapter/${i}  ${i}
done

# --minimum-length å¦‚æœå‰”é™¤æ¥å¤´åreadé•¿åº¦ä½äº30ï¼Œè¿™æ¡readå°†ä¼šè¢«ä¸¢å¼ƒ
# --overlap        å¦‚æœä¸¤ç«¯çš„åºåˆ—ä¸æ¥å¤´æœ‰4ä¸ªç¢±åŸºçš„åŒ¹é…å°†ä¼šè¢«å‰”é™¤
# --trim-n         å‰”é™¤ä¸¤ç«¯çš„N
# -a å»é™¤3ç«¯å¼•ç‰©åºåˆ—
```
## 4.3 å†æ¬¡å»é™¤ä½è´¨é‡åŒºåŸŸ
```
parallel -j 4 "
java -jar /mnt/c/project/biosoft/Trimmomatic-0.38/Trimmomatic-0.38.jar \                                                       SE -phred33 {1} ../trim/{1} \
LEADING:20 TRAILING:20 SLIDINGWINDOW:5:15 MINLEN:30 \
" ::: $( ls *.gz)

# æœ¬å‘½ä»¤é€»è¾‘  
java -jar  [Trimmomaticè½¯ä»¶å­˜å‚¨ä½ç½®]\   #-jar æ‰§è¡Œå°è£…åœ¨ JAR å­˜æ¡£ä¸­çš„ç¨‹åº
   [å•ç«¯æµ‹åº] SE -phred33 {å˜é‡åç§°}      [æŒ‡æ˜å­˜å‚¨è·¯å¾„å’Œåç§°] ../trim/{å˜é‡åç§°}\
   Trimmomaticè½¯ä»¶çš„å‘½ä»¤é€‰é¡¹  


# LEADING:20ï¼Œä»åºåˆ—çš„å¼€å¤´å¼€å§‹å»æ‰è´¨é‡å€¼å°äº 20 çš„ç¢±åŸº
# TRAILING:20ï¼Œä»åºåˆ—çš„æœ«å°¾å¼€å§‹å»æ‰è´¨é‡å€¼å°äº 20 çš„ç¢±åŸº
# SLIDINGWINDOW:5:15ï¼Œä» 5' ç«¯å¼€å§‹ä»¥ 5bp çš„çª—å£è®¡ç®—ç¢±åŸºå¹³å‡è´¨é‡ï¼Œå¦‚æœæ­¤å¹³å‡å€¼ä½äº 15ï¼Œåˆ™ä»è¿™ä¸ªä½ç½®æˆªæ–­read
# MINLEN:30ï¼Œ å¦‚æœ reads é•¿åº¦å°äº 30bp åˆ™æ‰”æ‰æ•´æ¡ readã€‚


#æˆ–è€…è¿è¡Œä¸‹é¢çš„ä»£ç ä¹Ÿå¯ä»¥
parallel -j 4 " trimmomatic SE -phred33 {1} ../trim/{1}  LEADING:20 TRAILING:20 SLIDINGWINDOW:5:15 MINLEN:30 " :::$(ls *.gz)
```
ç»“æœï¼š # Trimmomaticè½¯ä»¶å¤„ç†è½¬å…¥åå°ï¼Œä¸ä¼šæ˜¾ç¤ºå®æ—¶æ­¥éª¤ï¼Œè€å¿ƒç­‰å¾…
## 4.4 å†æ¬¡æŸ¥çœ‹è´¨é‡æƒ…å†µ
```
$ cd /mnt/c/project/rat/output/trim
# æŒ‚è½½åˆ°å­˜å‚¨gzæ–‡ä»¶çš„æ–‡ä»¶å¤¹ä¸‹

$ mkdir ../fastqc_trim
$ parallel -j 4 "
    fastqc -t 4 -o ../fastqc_trim {1}   #ä¸å¯å°†1æ¢æˆi
" ::: $(ls *.gz)

$ cd ../fastqc_trim
$ multiqc .
```
å¯è§‚å¯Ÿåˆ°ï¼Œæ¯”åŸæ¥çš„æƒ…å†µå¥½äº†å¾ˆå¤š

# 6. åºåˆ—æ¯”å¯¹
## 6.1 å»ºç«‹ç´¢å¼•
ä½¿ç”¨hisat2ä¸­çš„å·¥å…·hisat2-buildå»ºç«‹ç´¢å¼•ã€‚

* hisat2-build ç”¨æ³•
```
hisat2-build [options]* <reference_in> <ht2_base>
hisat2-build [é€‰é¡¹] [åŸºå› ç»„åºåˆ—(.fa)] [ç´¢å¼•æ–‡ä»¶çš„å‰ç¼€å]

#<reference_in> ï¼šfastaæ–‡ä»¶;  å¦‚æœä¸ºlistï¼Œä½¿ç”¨é€—å·åˆ†å¼€
#<ht2_base> ï¼šç´¢å¼•æ–‡ä»¶çš„å‰ç¼€åï¼Œå¦‚è®¾ä¸ºxxxï¼Œåˆ™ç”Ÿæˆçš„ç´¢å¼•æ–‡ä»¶ä¸ºxxx.1.ht2,xxx.2.ht2ï¼Œé»˜è®¤çš„å‰ç¼€åä¸ºNAME

```
å¼€å§‹ä½¿ç”¨ ! å¸ˆå…„åªç”¨ä¸€å·æŸ“è‰²ä½“åšäº†æ¯”å¯¹ï¼Œå‘ç°ç»“æœä¸å¤ªå¯¹ï¼Œæ‰€ä»¥æ¥ä¸‹æ¥ä»¥å…¨åŸºå› ç»„åšæ¯”å¯¹
```
$ cd /mnt/c/project/rat/genome
$ mkdir index
$ cd index
 
hisat2-build  -p 6 ../mRatBN7.2.fa mRatBN7.2
#-p å¹¶è¡Œè¿ç®—çº¿ç¨‹æ•°ä¸º6ï¼Œ-på¹¶è¡Œè¿ç®—çº¿ç¨‹æ•°(é»˜è®¤:1)ã€‚
```
åœ¨è¿è¡Œè¿‡ç¨‹ä¸­ä¼šæœ‰éƒ¨åˆ†ä¿¡æ¯æç¤ºï¼Œå…¶ä¸­è¯´åˆ°å»ºç«‹ç´¢å¼•æ–‡ä»¶çš„åˆ†å—æƒ…å†µä»¥åŠè¿è¡Œæ—¶é—´çš„ç»Ÿè®¡
ç´¢å¼•å»ºç«‹å®Œæˆä¹‹ååœ¨c:\project\rat\genome\indexæ–‡ä»¶å¤¹ä¸‹ä¼šå‡ºç°
```
mRatBN7.2.1.ht2
mRatBN7.2.2.ht2
...
mRatBN7.2.8.ht2
```
8ä¸ªæ–‡ä»¶ï¼Œè¿™äº›æ–‡ä»¶æ˜¯å¯¹åŸºå› ç»„è¿›è¡Œå‹ç¼©ä¹‹åçš„æ–‡ä»¶ï¼Œè¿™ä¸ªå°†åŸºå› ç»„åºåˆ—æ•°æ®åˆ†å—æˆäº†8ä»½ï¼Œåœ¨æ‰§è¡Œåºåˆ—æ¯”å¯¹çš„æ—¶å€™ç›´æ¥ä½¿ç”¨è¿™äº›æ–‡ä»¶è€Œä¸æ˜¯åŸºå› ç»„mRatBN7.2.faæ–‡ä»¶ã€‚

## 6.2 å¼€å§‹æ¯”å¯¹
è¿™é‡Œä½¿ç”¨hisat2è¿›è¡Œæ¯”å¯¹
```
Usage:
  hisat2 [options]* -x <ht2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]

  <ht2-idx>  Index filename prefix (minus trailing .X.ht2).
  <m1>       Files with #1 mates, paired with files in <m2>.
             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).
  <m2>       Files with #2 mates, paired with files in <m1>.
             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).
  <r>        Files with unpaired reads.
             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).
  <sam>      File for SAM output (default: stdout)
```

```
# åŸæ¥çš„ä»£ç 
parallel -k -j 4 "    #-k --keep-order å¼ºåˆ¶ä½¿è¾“å‡ºä¸å‚æ•°ä¿æŒé¡ºåºï¼›-j å¹¶è¡Œä»»åŠ¡æ•°
    hisat2 -t -x ../../genome/index/mRatBN7.2 \    # -x å®šä½åˆ°ç´¢å¼•æ–‡ä»¶
      -U {1}.fastq.gz -S ../align/{1}.sam \
      # -U <r>ï¼šå•ç«¯æµ‹åºlistï¼Œè‹¥ä¸ºlistï¼Œä½¿ç”¨é€—å·éš”å¼€ï¼Œ-U lane1.fq,lane2.fq,lane3.fq,lane4.fq
      # -S <hit> ï¼šSAMå†™å…¥çš„æ–‡ä»¶åï¼Œé»˜è®¤å†™å…¥åˆ°æ ‡å‡†è¾“å‡ºä¸­

      2>../align/{1}.log      # æ ‡å‡†è¾“å‡ºè¾“å…¥åˆ°ç›¸å¯¹åº”çš„æ–‡ä»¶ä¸­
" ::: $(ls *.gz | perl -p -e 's/.fastq.gz$//')    # perl è¯­å¥èµ·åˆ°æ‰¹é‡æ›¿æ¢ä½œç”¨ï¼Œå°†æ–‡ä»¶æ ¼å¼åç¼€å»æ‰


# æŠŠç´¢å¼•å’Œæµ‹åºæ•°æ®æ”¾åˆ°åŒä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œé¢
hisat2 -t -x mRatBN7.2 -U SRR2190795.fastq.gz -S SRR2190795.sam

parallel -k -j 2 " 
    hisat2 -t -x mRatBN7.2 \
      -U {1}.fastq.gz -S {1}.sam \
      " ::: $(ls *.gz | perl -p -e 's/.fastq.gz$//')
```


## æ­£ç¡®çš„åšæ³•

* forå¾ªç¯

ç”±äºå†…å­˜å ç”¨å¤ªå¤§ï¼Œå æ»¡äº†ï¼Œç¨‹åºè·‘ä¸èµ·æ¥ï¼Œæ‰€ä»¥å´å¸ˆå…„å¸®å¿™å†™äº†ä¸€ä¸ª``forå¾ªç¯``ï¼Œç¨‹åºæ­£å¸¸è¿è¡Œã€‚æˆ‘æŠŠindexç´¢å¼•æ–‡ä»¶ï¼Œè¾“å…¥æ–‡ä»¶å’Œè¾“å‡ºæ–‡ä»¶æ”¾åˆ°äº†åŒä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œé¢ã€‚ï¼ˆçœ‹åˆ°ç½‘ä¸Šè¯´åœ¨åŒä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œé¢æŠ¥é”™çš„å¯èƒ½æ€§å°ä¸€äº›ã€‚ï¼‰
```
guoqinghua@DESKTOP-VPR0E67:/mnt/c/project/rat/output/trim$ for i in $(ls *.gz | perl -p -e 's/.fastq.gz$//')
> do
> hisat2 -t -x mRatBN7.2 -U ${i}.fastq.gz -S ${i}.sam 2>../align/${i}.log
> done
```
* ä¸€ä¸ªä¸€ä¸ªå•ç‹¬è¿è¡Œã€‚ç´¢å¼•æ–‡ä»¶å’Œè¾“å…¥è¾“å‡ºæ–‡ä»¶ä¸åœ¨åŒä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œé¢ã€‚ç•Œé¢è¾“å‡ºä¿¡æ¯å†™å…¥åˆ°æ—¥å¿—ä¸­
```
guoqinghua@DESKTOP-VPR0E67:/mnt/c/project/rat/output/trim$ hisat2 -t -x ../../genome/index/mRatBN7.2 -U SRR2190795.fastq.gz -S ../align/SRR2190795.sam 2>../align/SRR2190795.log
```
* 3.ä¸€ä¸ªä¸€ä¸ªå•ç‹¬è¿è¡Œã€‚
ç´¢å¼•æ–‡ä»¶å’Œè¾“å…¥è¾“å‡ºæ–‡ä»¶ä¸åœ¨åŒä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œé¢ã€‚ç•Œé¢è¾“å‡ºä¿¡æ¯æœªå†™å…¥åˆ°æ—¥å¿—ä¸­ï¼Œç¨‹åºè·‘å®Œä¹‹ååœ¨ç»ˆç«¯ç•Œé¢æ˜¾ç¤ºã€‚
```
guoqinghua@DESKTOP-VPR0E67:/mnt/c/project/rat/output/trim$ hisat2 -t -x ../../genome/index/mRatBN7.2 -U SRR2240187.fastq.gz -S ../align/SRR2240187.sam

Time loading forward index: 00:01:51
Time loading reference: 00:00:03
Multiseed full-index search: 00:36:11
19363668 reads; of these:
  19363668 (100.00%) were unpaired; of these:
    1444081 (7.46%) aligned 0 times
    16869438 (87.12%) aligned exactly 1 time
    1050149 (5.42%) aligned >1 times
92.54% overall alignment rate
Time searching: 00:36:15
Overall time: 00:38:06
```
## 6.3 æ ¼å¼è½¬åŒ–ä¸æ’åº
åŸä»£ç 
```
parallel -k -j 4 "
    samtools sort -@ 4 {1}.sam > {1}.sort.bam
    samtools index {1}.sort.bam
" ::: $(ls *.sam | perl -p -e 's/\.sam$//')

rm *.sam
```
æˆ‘è¾“å…¥çš„ä»£ç 
```
parallel -k -j 1 "
    samtools sort -@ 1 {1}.sam > {1}.sort.bam
" ::: $(ls *.sam | perl -p -e 's/\.sam$//')

# æœ‰ä¸€ä¸ªæ²¡æœ‰è·‘æˆåŠŸï¼Œæ‰€ä»¥å•ç‹¬è¿è¡Œä¸€æ¬¡
samtools sort SRR2240186.sam > SRR2240186.sort.bam 
```
* samtools sort
```
Usage: samtools sort [options...] [in.bam]
options:
-n : æ ¹æ®readçš„nameè¿›è¡Œæ’åºï¼Œé»˜è®¤å¯¹æœ€å·¦ä¾§åæ ‡è¿›è¡Œæ’åº
-o : è®¾ç½®æ’åºåè¾“å‡ºæ–‡ä»¶çš„æ–‡ä»¶å
-O : åè·Ÿsamæˆ–bamï¼Œè§„å®šæ’åºåè¾“å‡ºæ–‡ä»¶çš„æ ¼å¼ï¼Œé»˜è®¤æ˜¯bam
-@ : åè·Ÿæ­£æ•´æ•°ï¼ŒæŒ‡å®šåˆ†ææ‰€ç”¨çº¿ç¨‹æ•°
-m : åè·Ÿæ•°å­— + K/M/Gï¼ŒæŒ‡å®šæ¯ä¸ªçº¿ç¨‹æ‰€ä½¿ç”¨å†…å­˜é‡  
```
* samtools index  
```
 ä¸ºäº†èƒ½å¤Ÿå¿«é€Ÿè®¿é—®bamæ–‡ä»¶ï¼Œå¯ä»¥ä¸ºå·²ç»åŸºäºåæ ‡æ’åºåbamæˆ–è€…cramçš„æ–‡ä»¶åˆ›å»ºç´¢å¼•ï¼Œç”Ÿæˆä»¥.baiæˆ–è€….craiä¸ºåç¼€çš„ç´¢å¼•æ–‡ä»¶ã€‚
 å¿…é¡»ä½¿ç”¨æ’åºåçš„æ–‡ä»¶ï¼Œå¦åˆ™å¯èƒ½ä¼šæŠ¥é”™ã€‚
 å¦å¤–ï¼Œä¸èƒ½å¯¹samæ–‡ä»¶ä½¿ç”¨æ­¤å‘½ä»¤ã€‚å¦‚æœæƒ³å¯¹samæ–‡ä»¶å»ºç«‹ç´¢å¼•ï¼Œé‚£ä¹ˆå¯ä»¥ä½¿ç”¨tabixå‘½ä»¤åˆ›å»ºã€‚

indexå‘½ä»¤æ ¼å¼å¦‚ä¸‹ï¼š
      samtools index [-bc] [-m INT] aln.bam |aln.cram [out.index]
      å‚æ•°ï¼š
      -b åˆ›å»ºbaiç´¢å¼•æ–‡ä»¶ï¼ŒæœªæŒ‡å®šè¾“å‡ºæ ¼å¼æ—¶ï¼Œæ­¤å‚æ•°ä¸ºé»˜è®¤å‚æ•°ï¼›
      -c åˆ›å»ºcsiç´¢å¼•æ–‡ä»¶ï¼Œé»˜è®¤æƒ…å†µä¸‹ï¼Œç´¢å¼•çš„æœ€å°é—´éš”å€¼ä¸º2^14ï¼Œä¸baiæ ¼å¼ä¸€è‡´ï¼›
      -m INT åˆ›å»ºcsiç´¢å¼•æ–‡ä»¶ï¼Œæœ€å°é—´éš”å€¼2^INTï¼›
```
è¾“å…¥ä»£ç ï¼š
```
for i in $(ls *.sort.bam)
   do
   echo ${i}
   samtools index ${i}
   done
```

### 7. è¡¨è¾¾é‡ç»Ÿè®¡
```
cd /mnt/c/project/rat/annotation

# ä¸‹è½½.gtfæ³¨é‡Šæ–‡ä»¶
wget http://ftp.ensembl.org/pub/release-107/gff3/rattus_norvegicus/Rattus_norvegicus.mRatBN7.2.107.gff3.gz
```

```
cd /mnt/c/project/rat/output/align
parallel -j 4 "
 htseq-count -s no -f bam {1}.sort.bam ../../annotation/mRatBN7.2.107.gtf \
 >../HTseq/{1}.count 2>../HTseq/{1}.log
 " ::: $(ls *.sort.bam | perl -p -e 's/\.sort\.bam$//')
``` 
```
cd /mnt/c/project/rat/output/HTseq
cat SRR2190795.fastq.gz.count | head -n 10
ENSRNOG00000000001      2
ENSRNOG00000000007      3
ENSRNOG00000000008      0
ENSRNOG00000000009      0
ENSRNOG00000000010      0
ENSRNOG00000000012      0
ENSRNOG00000000017      10
ENSRNOG00000000021      0
ENSRNOG00000000024      844
ENSRNOG00000000033      28
```
### 8. åˆå¹¶è¡¨è¾¾çŸ©é˜µä¸æ ‡å‡†åŒ–
#### 8.1 åˆå¹¶

ä¸‹é¢ä½¿ç”¨Rè¯­è¨€ä¸­çš„mergeå°†è¡¨æ ¼åˆå¹¶
```
# å…ˆè¾“å…¥å¤§å†™Rï¼Œå†è¾“å…¥ä¸‹åˆ—å‘½ä»¤ï¼›æˆ–ä½¿ç”¨Rstudioï¼Œåº”ä¿è¯è·¯å¾„ä¸€è‡´ï¼Œç›®å‰è¿˜ä¸å¤ªä¼š
rm(list=ls())
setwd("/mnt/d/project/rat/output/HTseq")

# å¾—åˆ°æ–‡ä»¶æ ·æœ¬ç¼–å·
files <- list.files(".", "*.count")
f_lists <- list()
for(i in files){
    prefix = gsub("(_\\w+)?\\.count", "", i, perl=TRUE)
    f_lists[[prefix]] = i
}

id_list <- names(f_lists)
data <- list()
count <- 0
for(i in id_list){
  count <- count + 1
  a <- read.table(f_lists[[i]], sep="\t", col.names = c("gene_id",i))
  data[[count]] <- a
}

# åˆå¹¶æ–‡ä»¶
data_merge <- data[[1]]
for(i in seq(2, length(id_list))){
    data_merge <- merge(data_merge, data[[i]],by="gene_id")
}

write.csv(data_merge, "merge.csv", quote = FALSE, row.names = FALSE)
```
è¡¨æ ¼åœ¨HTseqæ–‡ä»¶å¤¹ä¸­
#### 8.2 æ•°æ®æ ‡å‡†åŒ–

å®‰è£…RåŒ…"GenomicFeatures",åœ¨ubuntuä¸­è¾“å…¥
```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("GenomicFeatures")
```
é€‰æ‹©é•œåƒï¼Œä»»æ„é€‰æ‹©ä¸€ä¸ªä¸­å›½çš„éƒ½å¯ä»¥ï¼Œè¿™é‡Œæˆ‘é€‰çš„ nju.edu.cn
å®‰è£…Rç¨‹åºåŒ…æ—¶ï¼Œå‡ºç°äº†ä¸€äº›é—®é¢˜ï¼Œ
ç„¶åä»¥ç®¡ç†å‘˜èº«ä»½æ‰“å¼€Rstudioï¼Œ
```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("GenomicFeatures")
```
æŠ¥é”™æç¤º
```
Warning message:
package(s) not installed when version(s)
  same as or greater than current; use
  `force = TRUE` to re-install:
  'GenomicFeatures'
```
é‡æ–°è¾“å…¥
```
 BiocManager::install("GenomicFeatures"ï¼Œforce = TRUE)
```
ä¾æ—§ä¸èƒ½æ­£å¸¸å®‰è£…ã€‚
éœ€è¦å®‰è£…å…¶å®ƒçš„package
```
>library(GenomicFeatures)
Error: package or namespace load failed for â€˜GenomicFeaturesâ€™ in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
 ä¸å­˜åœ¨å«â€˜prettyunitsâ€™è¿™ä¸ªåå­—çš„ç¨‹è¾‘åŒ…

>BiocManager::install("prettyunits")
...
Error in download.file(url, destfile, method, mode = "wb", ...) : 
  cannot open URL 'https://cran.rstudio.com/bin/windows/contrib/4.3/prettyunits_1.1.1.zip'
In addition: Warning message:
In download.file(url, destfile, method, mode = "wb", ...) :
  URL 'https://cran.rstudio.com/bin/windows/contrib/4.3/prettyunits_1.1.1.zip': status was 'SSL connect error'
Warning in download.packages(pkgs, destdir = tmpd, available = available,  :
  ä¸‹è½½ç¨‹åºåŒ…â€˜prettyunitsâ€™æ—¶å‡ºäº†é—®é¢˜
```
å¯ä»¥å‚è€ƒä»¥ä¸‹ç½‘å€ï¼Œä¿®æ”¹é…ç½®ã€‚
Win+Rï¼Œè¾“å…¥inetcpl.cpl ç›´æ¥æ‰“å¼€Interneté€‰é¡¹ã€‚æ‰“å¼€åï¼Œåœ¨é«˜çº§ä¸­å‹¾é€‰ä½¿ç”¨TLS 1.0ã€ä½¿ç”¨TLS 1.1ã€ä½¿ç”¨TLS 1.2ã€‚

ä½†å…¶å®ä¿®æ”¹å®Œï¼Œå…³é—­é˜²ç«å¢™ä¹‹åæœ‰æ—¶ä¹Ÿè¿˜æ˜¯ä¸è¡Œï¼Œå¤šå®‰è£…å‡ æ¬¡å°±å¯ä»¥å®‰è£…ä¸Šäº†ï¼Œå¯èƒ½æ˜¯ç½‘ç»œä¸ç¨³å®šé—®é¢˜ã€‚

å‚è€ƒç½‘å€ï¼š

https://www.cnblogs.com/miyuanbiotech/p/14077003.html

https://www.zhihu.com/question/48323000

Rè¯­è¨€ä¸­è®¾ç½®å·¥ä½œè·¯å¾„
```
setwd("C:/project/rat/annotation")
```
txdb <- makeTxDbFromGFF("mRatBN7.2.107.gtf",format="gtf")


```
# å…ˆä¸‹è½½ GenomicFeatures
BiocManager::install('GenomicFeatures')
library(GenomicFeatures)

# æ„å»ºGrangeså¯¹è±¡
txdb <- makeTxDbFromGFF("mRatBN7.2.gff3",format="gff3")
# æˆ–è€…ç”¨gtfå°è¯•ä¸€ä¸‹ï¼Œæœ‰ä»€ä¹ˆåŒºåˆ«
# txdb <- makeTxDbFromGFF("mRatBN7.2.107.gtf",format="gtf")
#Warning message:
  In .get_cds_IDX(mcols0$type, mcols0$phase) :
  The "phase" metadata column contains non-NA values for features of type
  stop_codon. This information was ignored.

# æŸ¥æ‰¾åŸºå› çš„å¤–æ˜¾å­
exons_gene <- exonsBy(txdb, by = "gene")

# è®¡ç®—æ€»é•¿åº¦
# reduce()ã€width()æ˜¯Irangeå¯¹è±¡çš„æ–¹æ³•
gene_len <- list()
for(i in names(exons_gene)){
    range_info = reduce(exons_gene[[i]])
    width_info = width(range_info)
    sum_len    = sum(width_info)
    gene_len[[i]] = sum_len
}

# æˆ–è€…å†™ä¸ºlapplyçš„å½¢å¼(å¿«å¾ˆå¤š)
gene_len <- lapply(exons_gene,function(x){sum(width(reduce(x)))})

#æŸ¥çœ‹ä¸‹
class(gene_len)
length(gene_len)
```





```
å…ˆæŠŠmRatBN7.2_gene_len.tsv"
#å°†åŸæ¥å‚¨å­˜åœ¨annotationä¸­çš„å¤–æ˜¾å­é•¿åº¦æ–‡ä»¶ç²˜è´´åˆ°count æ‰€åœ¨çš„ HTseq æ–‡ä»¶å¤¹ä¸‹ 
rm(list=ls())
> setwd("C:/project/rat/output/HTseq")
> gene_len_file <- "mRatBN7.2_gene_len.tsv"
> count_file <- "SRR2190795.count"
> gene_len <- read.table(gene_len_file, header = FALSE, row.name = 1)
> colnames(gene_len) <- c("length")
> count <- read.table(count_file, header = FALSE, row.name = 1)
> colnames(count) <- c("count")
> all_count <- sum(count["count"])
> RPKM <- c()
> for(i in row.names(count)){
+ count_ = 0
+ exon_kb = 1
+ rpkm = 0
+ count_ = count[i, ]
+ exon_kb  = gene_len[i, ] / 1000
+ rpkm    = (10 ^ 6 * count_ ) / (exon_kb * all_count )
+ RPKM = c(RPKM, rpkm)
+ }
```


```
setwd("C:/project/rat/annotation")
library(GenomicFeatures)
txdb <- makeTxDbFromGFF("mRatBN7.2.gff3",format="gff3")
exons_gene <- exonsBy(txdb, by = "gene")

gene_len <- list()
for(i in names(exons_gene)){
range_info = reduce(exons_gene[[i]])
width_info = width(range_info)
sum_len    = sum(width_info)
gene_len[[i]] = sum_len
}
# æˆ–è€…å†™ä¸ºlapplyçš„å½¢å¼(å¿«å¾ˆå¤š)
gene_len <- lapply(exons_gene,function(x){sum(width(reduce(x)))})

length(gene_len)
class(gene_len)
data <- t(as.data.frame(gene_len))
write.table(data, file = "mRatBN7.2_gene_len.tsv", row.names = TRUE, sep="\t", quote = FALSE, col.names = FALSE)
```